pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,858 @@
+		}
+		filter := bpf.CompilePacketFilter(pc.Spec.Packet, matchPacket)
+		klog.V(5).InfoS("Generated bpf instructions for Packetcapture", "name", pc.Name, "match packet", matchPacket, "bpf instructions", filter)
+		return NewPcapSource(*device, filter)
+	}
+
+	return c
+}
+
+func (c *Controller) enqueuePacketCapture(pc *crdv1alpha1.PacketCapture) {
+	c.queue.Add(pc.Name)
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the PacketCapture events from the
+// workqueue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting packetcapture controller", "name", controllerName)
+	defer klog.InfoS("Shutting down packetcapture controller", "name", controllerName)
+
+	cacheSynced := []cache.InformerSynced{c.packetCaptureSynced}
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSynced...) {
+		return
+	}
+
+	err := defaultFS.MkdirAll(packetDirectory, 0755)
+	if err != nil {
+		klog.ErrorS(err, "Couldn't create the directory for storing captured packets", "directory", packetDirectory)
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addPacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture ADD event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCapture(_, obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture UPDATE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	// check the captures that are waiting in line.
+	go c.processWaitingCaptures()
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processWaitingCaptures() {
+	for {
+		capture := func() string {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			for c.numRunningCaptures >= maxConcurrentCaptures || c.waitingCaptures == 0 {
+				c.cond.Wait()
+			}
+			c.numRunningCaptures += 1
+			c.waitingCaptures -= 1
+			capture := <-c.waitingCapturesCh
+			c.startPacketCapture(capture)
+			c.captures[capture] = packetCapturePhaseRunning
+			return capture
+		}()
+		c.queue.Add(capture)
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		if !apierrors.IsNotFound(err) {
+			c.queue.AddRateLimited(key)
+			klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+		} else {
+			c.queue.Forget(key)
+			delete(c.captures, key)
```

Comment:
The program would panic due to the data race.

And I don't think it should add and update items to the map in `syncPacketCapture` but delete items from the map outside `syncPacketCapture`

Reply:
updated. moved to `syncPacketCapture`.